Add priorities to policies

fredreichbier opened the pull request:
This PR
* adds a ``priority`` field to the policy table
* modifies ``PolicyObject.get_action_values`` to use priorities to resolve conflicts in case ``unique=True`` is passed
* modifies some policy decorators to either use ``get_action_values(unique=True)`` or to postprocess the output of ``get_policies``
* adds the priority field to the WebUI, documentation and translations

Some notes:
* Initially I thought ``PolicyObject.get_policies`` should just return the policies sorted by priority. However, I decided against this because this would change the order of policies *everywhere*, e.g. also in the policy list WebUI. So instead, the priority field is only processed in situations where policy conflicts can occur.
* ``choose_prioritized_policy`` inspects the given list and raises an error if there are any two policies with the same priority. This isn't really necessary: Having three matching policies ``pol1`` (with priority 1), ``pol2`` (prio 2) and ``pol3`` (prio 2) raises an error, but would be OK, as ``pol1`` is selected anyway.  However, I thought it is better to be safe than sorry, because multiple matching policies with the same priority hint at a configuration error :)
* This PR slightly changes behavior in some cases: In privacyIDEA 2.22, it is OK to have multiple matching policies ``otppin`` as long as they contain the same action value. With these changes, this would raise an error if they all have the same priority, even if they specify the same action value. Do you think we should keep backwards-compatibility here?
* The migration script uses ``server_default`` to set the priority of all existing policies to 1. This seems to work fine, though we could also use a separate ``UPDATE`` step instead.

I'll be happy for reviews (especially the changed policy decorators) and of course, I'm open to questions and suggestions :)

Closes #1031
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23issuecomment-388336853%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23issuecomment-388789683%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r187925004%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r187926517%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r187926664%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r187928375%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23issuecomment-388869718%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r188183197%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r188183584%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r188183959%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23issuecomment-388336853%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%23%20%5BCodecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062%3Fsrc%3Dpr%26el%3Dh1%29%20Report%5Cn%3E%20Merging%20%5B%231062%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062%3Fsrc%3Dpr%26el%3Ddesc%29%20into%20%5Bmaster%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/commit/93b86b9b1a886978771a452bc8d47cbae281ea6c%3Fsrc%3Dpr%26el%3Ddesc%29%20will%20%2A%2Aincrease%2A%2A%20coverage%20by%20%600.18%25%60.%5Cn%3E%20The%20diff%20coverage%20is%20%60100%25%60.%5Cn%5Cn%5B%21%5BImpacted%20file%20tree%20graph%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/graphs/tree.svg%3Fwidth%3D650%26src%3Dpr%26token%3D7nHLZki60B%26height%3D150%29%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062%3Fsrc%3Dpr%26el%3Dtree%29%5Cn%5Cn%60%60%60diff%5Cn%40%40%20%20%20%20%20%20%20%20%20%20%20%20Coverage%20Diff%20%20%20%20%20%20%20%20%20%20%20%20%20%40%40%5Cn%23%23%20%20%20%20%20%20%20%20%20%20%20master%20%20%20%20%231062%20%20%20%20%20%20%2B/-%20%20%20%23%23%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%2B%20Coverage%20%20%2095.43%25%20%20%2095.62%25%20%20%20%2B0.18%25%20%20%20%20%20%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%20%20Files%20%20%20%20%20%20%20%20%20131%20%20%20%20%20%20131%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5Cn%20%20Lines%20%20%20%20%20%20%2016239%20%20%20%2016804%20%20%20%20%20%2B565%20%20%20%20%20%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%2B%20Hits%20%20%20%20%20%20%20%2015498%20%20%20%2016068%20%20%20%20%20%2B570%20%20%20%20%20%5Cn%2B%20Misses%20%20%20%20%20%20%20%20741%20%20%20%20%20%20736%20%20%20%20%20%20%20-5%5Cn%60%60%60%5Cn%5Cn%5Cn%7C%20%5BImpacted%20Files%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062%3Fsrc%3Dpr%26el%3Dtree%29%20%7C%20Coverage%20%5Cu0394%20%7C%20%7C%5Cn%7C---%7C---%7C---%7C%5Cn%7C%20%5Bprivacyidea/api/lib/postpolicy.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvYXBpL2xpYi9wb3N0cG9saWN5LnB5%29%20%7C%20%6096.61%25%20%3C%5Cu00f8%3E%20%28%2B0.31%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%7C%20%5Bprivacyidea/models.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbW9kZWxzLnB5%29%20%7C%20%6099.12%25%20%3C100%25%3E%20%28%2B0.35%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%7C%20%5Bprivacyidea/lib/policydecorators.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL3BvbGljeWRlY29yYXRvcnMucHk%3D%29%20%7C%20%60100%25%20%3C100%25%3E%20%28%2B2.07%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%7C%20%5Bprivacyidea/lib/policy.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL3BvbGljeS5weQ%3D%3D%29%20%7C%20%6099.36%25%20%3C100%25%3E%20%28%2B0.01%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%7C%20%5Bprivacyidea/api/policy.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvYXBpL3BvbGljeS5weQ%3D%3D%29%20%7C%20%60100%25%20%3C100%25%3E%20%28%5Cu00f8%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%7C%20%5Bprivacyidea/lib/eventhandler/counterhandler.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL2V2ZW50aGFuZGxlci9jb3VudGVyaGFuZGxlci5weQ%3D%3D%29%20%7C%20%6095.45%25%20%3C0%25%3E%20%28-4.55%25%29%60%20%7C%20%3Aarrow_down%3A%20%7C%5Cn%7C%20%5Bprivacyidea/lib/counter.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL2NvdW50ZXIucHk%3D%29%20%7C%20%60100%25%20%3C0%25%3E%20%28%5Cu00f8%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%7C%20%5Bprivacyidea/lib/eventhandler/usernotification.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL2V2ZW50aGFuZGxlci91c2Vybm90aWZpY2F0aW9uLnB5%29%20%7C%20%6093.86%25%20%3C0%25%3E%20%28%2B3.43%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%5Cn------%5Cn%5Cn%5BContinue%20to%20review%20full%20report%20at%20Codecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062%3Fsrc%3Dpr%26el%3Dcontinue%29.%5Cn%3E%20%2A%2ALegend%2A%2A%20-%20%5BClick%20here%20to%20learn%20more%5D%28https%3A//docs.codecov.io/docs/codecov-delta%29%5Cn%3E%20%60%5Cu0394%20%3D%20absolute%20%3Crelative%3E%20%28impact%29%60%2C%20%60%5Cu00f8%20%3D%20not%20affected%60%2C%20%60%3F%20%3D%20missing%20data%60%5Cn%3E%20Powered%20by%20%5BCodecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062%3Fsrc%3Dpr%26el%3Dfooter%29.%20Last%20update%20%5B93b86b9...e900fb6%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1062%3Fsrc%3Dpr%26el%3Dlastupdated%29.%20Read%20the%20%5Bcomment%20docs%5D%28https%3A//docs.codecov.io/docs/pull-request-comments%29.%5Cn%22%2C%20%22created_at%22%3A%20%222018-05-11T11%3A24%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/8655789%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/codecov-io%22%7D%7D%2C%20%7B%22body%22%3A%20%22%5Cr%5CnSome%20thoughts%20on%20your%20comments%20before%20actually%20looking%20at%20the%20code%3A%5Cr%5Cn%5Cr%5Cn%3E%20Initially%20I%20thought%20PolicyObject.get_policies%20should%20just%20return%20the%20policies%20sorted%20by%20priority.%20However%2C%20I%20decided%20against%20this%20because%20this%20would%20change%20the%20order%20of%20policies%20everywhere%2C%20e.g.%20also%20in%20the%20policy%20list%20WebUI.%20So%20instead%2C%20the%20priority%20field%20is%20only%20processed%20in%20situations%20where%20policy%20conflicts%20can%20occur.%5Cr%5Cn%5Cr%5CnAcutally%20changing%20the%20order%20everywhere%20might%20be%20quite%20nice%2C%20wouldn%27t%20it%3F%20This%20way%20the%20admin%20could%20also%20%5C%22sort%5C%22%20the%20policies%20in%20the%20webui%2C%20which%20would%20also%20help%20him%20to%20understand%20the%20application%20of%20the%20policies.%5Cr%5Cn%5Cr%5Cn%3E%20This%20PR%20slightly%20changes%20behavior%20in%20some%20cases%3A%20In%20privacyIDEA%202.22%2C%20it%20is%20OK%20to%20have%20multiple%20matching%20policies%20otppin%20as%20long%20as%20they%20contain%20the%20same%20action%20value.%20With%20these%20changes%2C%20this%20would%20raise%20an%20error%20if%20they%20all%20have%20the%20same%20priority%2C%20even%20if%20they%20specify%20the%20same%20action%20value.%20Do%20you%20think%20we%20should%20keep%20backwards-compatibility%20here%3F%5Cr%5Cn%5Cr%5CnI%20think%20we%20should%20keep%20backwards-compatibility.%20%2Aotppin%2A%20is%20I%20think%20fetched%20by%20%60%60get_action_value%60%60.%20We%20could%20simply%20return%20the%20list%20of%20the%20highest%20priority%20action%20value.%20If%20there%20are%20more%20than%20one%20values/policies%20with%20different%20priority%2C%20we%20can%20still%20raise%20an%20exception.%22%2C%20%22created_at%22%3A%20%222018-05-14T11%3A49%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20added%20the%20priorities%20and%20the%20sorting%20functionality%20to%20the%20policy%20list%20and%20changed%20the%20conflict%20resolution%20as%20we%20discussed%20%28see%20%60%60get_action_values%60%60%29.%5Cr%5Cn%5Cr%5CnI%20also%20added%20the%20%60%60sort_by_priority%60%60%20flag%20to%20%60%60get_policies%60%60%20as%20discussed%20and%20set%20its%20default%20value%20to%20True.%20Thus%2C%20all%20policies%20are%20now%20sorted%20by%20priority.%20Doesn%27t%20seem%20to%20break%20anything.%20%3A%29%5Cr%5Cn%5Cr%5CnI%20am%20hesitant%20to%20add%20the%20%60%60unique%60%60%20flag%20to%20%60%60get_policies%60%60%3A%20Actually%2C%20it%20seems%20like%20%60%60auth_user_passthru%60%60%20is%20the%20only%20place%20where%20the%20action%20values%20are%20significant%20and%20%60%60get_policies%60%60%20is%20used%20instead%20of%20%60%60get_action_values%60%60.%20We%20cannot%20just%20use%20%60%60get_action_values%60%60%20there%20because%201%29%20%60%60get_action_values%60%60%20does%20not%20support%20boolean%20actions%20for%20now%20and%202%29%20passthru%20needs%20the%20name%20of%20the%20matching%20policy.%20So%20instead%20I%20just%20added%20a%20conflict%20checking%20function%20to%20%60%60PolicyObject%60%60%20which%20is%20then%20invoked%20by%20%60%60auth_user_passthru%60%60.%20I%20wasn%27t%20sure%20if%20one%20occurrence%20would%20justify%20adding%20just%20another%20functionality%20to%20%60%60get_policies%60%60.%5Cr%5CnWhat%20do%20you%20think%3F%22%2C%20%22created_at%22%3A%20%222018-05-14T15%3A59%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/28243%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/fredreichbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%201b4377383a712e33c51815887e5afceda22e8059%20privacyidea/lib/policy.py%2057%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r188183197%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Can%20we%20move%20this%20up%3F%20If%20there%20is%20only%20one%20policy%20we%20also%20do%20not%20need%20to%20create%20prioritizes%20policy...%5Cr%5Cn%5Cr%5CnSo%20we%20could%20do%20in%20line%20613%3A%5Cr%5Cn%5Cr%5Cn%20%20%20if%20policies%20and%20len%28policies%29%20%3E%201%3A%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222018-05-15T07%3A04%3A49Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/policy.py%3AL591-626%22%7D%2C%20%22Pull%20e900fb6ace72202a32a2b541120a1deea8e0efdd%20privacyidea/lib/policydecorators.py%2017%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r187926664%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22We%20then%20could%20also%20pass%20the%20%2Aunique%2A%20to%20the%20get_policies%20in%20this%20function.%22%2C%20%22created_at%22%3A%20%222018-05-14T12%3A02%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/policydecorators.py%3AL358-376%22%7D%2C%20%22Pull%20e900fb6ace72202a32a2b541120a1deea8e0efdd%20privacyidea/static/components/config/views/config.policies.details.html%2017%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r187928375%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22In%20the%20event%20handler%20list%20we%20also%20have%20an%20input%20field%20to%20change%20the%20%5C%22ordering%5C%22%20%28priority%29%20of%20the%20event%20handlers%2C%20which%20is%20actually%20used%20to%20determine%2C%20in%20which%20order%20the%20event%20handlers%20are%20applied.%5Cr%5Cnhttps%3A//github.com/privacyidea/privacyidea/blob/master/privacyidea/static/components/config/views/config.events.list.html%23L19%5Cr%5Cn%5Cr%5CnI%20would%20love%20to%20be%20able%20to%20also%20change%20the%20same/similar%20way%20the%20priority%20in%20the%20policy%20list.%20This%20way%20the%20admin%20could%20%5C%22sort%5C%22%20the%20policy%20list%20and%20move%20policies%20up%20and%20down.%22%2C%20%22created_at%22%3A%20%222018-05-14T12%3A06%3A52Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/static/components/config/views/config.policies.details.html%3AL322-342%22%7D%2C%20%22Pull%201b4377383a712e33c51815887e5afceda22e8059%20privacyidea/lib/policy.py%20107%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r188183584%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20like%20this%21%20%3A-%29%22%2C%20%22created_at%22%3A%20%222018-05-15T07%3A06%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/policy.py%3AL671-682%22%7D%2C%20%22Pull%20e900fb6ace72202a32a2b541120a1deea8e0efdd%20privacyidea/lib/policy.py%2042%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r187925004%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20think%20this%20might%20often%20be%20a%20non-expected%20result.%20If%20there%20are%20two%20policies%2C%20with%20the%20same%20priority%20and%20the%20same%20resulting%20value%20-%20why%20should%20this%20raise%20an%20error%3F%5Cr%5Cn%5Cr%5CnThe%20admin%20would%20argue%2C%20that%20the%20configuration%20create%20a%20clear%20logic%20with%20no%20contradiction.%20So%20he%20probalby%20would%20expect%20privacyIDEA%20to%20act%20accordingly.%22%2C%20%22created_at%22%3A%20%222018-05-14T11%3A55%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/policy.py%3AL590-623%22%7D%2C%20%22Pull%20e900fb6ace72202a32a2b541120a1deea8e0efdd%20privacyidea/lib/policydecorators.py%207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r187926517%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Should%20we%20add%20this%20logic%20to%20%60%60get_policies%60%60%3F%5Cr%5Cn%5Cr%5CnAs%20we%20now%20have%20a%20solution%20for%20contradiction%20policies%2C%20there%20should%20be%20no%20contradicting%20policies%20anymore.%20So%20we%20could%20move%20the%20remaining%20contradicting%20stuff%20the%20the%20lower%20level%20of%20get_policies.%5Cr%5Cn%5Cr%5CnDepending%20on%20additional%20parameters%20like%20%2Asorted%2A%20and%20%2Aunique%2A%20we%20could%5Cr%5Cn%5Cr%5Cn1.%20return%20a%20sorted%20list%2C%20like%20for%20displaying%20in%20the%20webui.%5Cr%5Cn2.%20return%20the%20top%20most%20priority%20policy%20if%20%2Aunique%2A%20is%20set.%20And%20only%20if%20the%20logic%20is%20not%20clear%2C%20raise%20an%20exception%20in%20case%20of%20a%20failing%20%2Aunique%2A%20from%20within%20%60%60get_policies%60%60.%22%2C%20%22created_at%22%3A%20%222018-05-14T12%3A01%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/policydecorators.py%3AL300-308%22%7D%2C%20%22Pull%201b4377383a712e33c51815887e5afceda22e8059%20privacyidea/lib/policydecorators.py%2076%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1062%23discussion_r188183959%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20think%20we%20can%20combine%20this%20to%3A%5Cr%5Cn%5Cr%5Cn%7E%7E%7E%7Epython%5Cr%5Cnif%20login_mode_list%20and%20login_mode_list%5B0%5D%20%3D%3D%20LOGINMODE%3APRIVACYIDEA%3A%22%2C%20%22created_at%22%3A%20%222018-05-15T07%3A08%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/policydecorators.py%3AL515-526%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [x] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1062#issuecomment-388336853'>General Comment</a></b>
- <a href='https://github.com/codecov-io'><img border=0 src='https://avatars0.githubusercontent.com/u/8655789?v=4' height=16 width=16></a> # [Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1062?src=pr&el=h1) Report
> Merging [#1062](https://codecov.io/gh/privacyidea/privacyidea/pull/1062?src=pr&el=desc) into [master](https://codecov.io/gh/privacyidea/privacyidea/commit/93b86b9b1a886978771a452bc8d47cbae281ea6c?src=pr&el=desc) will **increase** coverage by `0.18%`.
> The diff coverage is `100%`.
[![Impacted file tree graph](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/graphs/tree.svg?width=650&src=pr&token=7nHLZki60B&height=150)](https://codecov.io/gh/privacyidea/privacyidea/pull/1062?src=pr&el=tree)
```diff
@@            Coverage Diff             @@
##           master    #1062      +/-   ##
==========================================
+ Coverage   95.43%   95.62%   +0.18%
==========================================
Files         131      131
Lines       16239    16804     +565
==========================================
+ Hits        15498    16068     +570
+ Misses        741      736       -5
```
| [Impacted Files](https://codecov.io/gh/privacyidea/privacyidea/pull/1062?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [privacyidea/api/lib/postpolicy.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvYXBpL2xpYi9wb3N0cG9saWN5LnB5) | `96.61% <ø> (+0.31%)` | :arrow_up: |
| [privacyidea/models.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbW9kZWxzLnB5) | `99.12% <100%> (+0.35%)` | :arrow_up: |
| [privacyidea/lib/policydecorators.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL3BvbGljeWRlY29yYXRvcnMucHk=) | `100% <100%> (+2.07%)` | :arrow_up: |
| [privacyidea/lib/policy.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL3BvbGljeS5weQ==) | `99.36% <100%> (+0.01%)` | :arrow_up: |
| [privacyidea/api/policy.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvYXBpL3BvbGljeS5weQ==) | `100% <100%> (ø)` | :arrow_up: |
| [privacyidea/lib/eventhandler/counterhandler.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL2V2ZW50aGFuZGxlci9jb3VudGVyaGFuZGxlci5weQ==) | `95.45% <0%> (-4.55%)` | :arrow_down: |
| [privacyidea/lib/counter.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL2NvdW50ZXIucHk=) | `100% <0%> (ø)` | :arrow_up: |
| [privacyidea/lib/eventhandler/usernotification.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1062/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL2V2ZW50aGFuZGxlci91c2Vybm90aWZpY2F0aW9uLnB5) | `93.86% <0%> (+3.43%)` | :arrow_up: |
------
[Continue to review full report at Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1062?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1062?src=pr&el=footer). Last update [93b86b9...e900fb6](https://codecov.io/gh/privacyidea/privacyidea/pull/1062?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> Some thoughts on your comments before actually looking at the code:
> Initially I thought PolicyObject.get_policies should just return the policies sorted by priority. However, I decided against this because this would change the order of policies everywhere, e.g. also in the policy list WebUI. So instead, the priority field is only processed in situations where policy conflicts can occur.
Acutally changing the order everywhere might be quite nice, wouldn't it? This way the admin could also "sort" the policies in the webui, which would also help him to understand the application of the policies.
> This PR slightly changes behavior in some cases: In privacyIDEA 2.22, it is OK to have multiple matching policies otppin as long as they contain the same action value. With these changes, this would raise an error if they all have the same priority, even if they specify the same action value. Do you think we should keep backwards-compatibility here?
I think we should keep backwards-compatibility. *otppin* is I think fetched by ``get_action_value``. We could simply return the list of the highest priority action value. If there are more than one values/policies with different priority, we can still raise an exception.
- <a href='https://github.com/fredreichbier'><img border=0 src='https://avatars2.githubusercontent.com/u/28243?v=4' height=16 width=16></a> I added the priorities and the sorting functionality to the policy list and changed the conflict resolution as we discussed (see ``get_action_values``).
I also added the ``sort_by_priority`` flag to ``get_policies`` as discussed and set its default value to True. Thus, all policies are now sorted by priority. Doesn't seem to break anything. :)
I am hesitant to add the ``unique`` flag to ``get_policies``: Actually, it seems like ``auth_user_passthru`` is the only place where the action values are significant and ``get_policies`` is used instead of ``get_action_values``. We cannot just use ``get_action_values`` there because 1) ``get_action_values`` does not support boolean actions for now and 2) passthru needs the name of the matching policy. So instead I just added a conflict checking function to ``PolicyObject`` which is then invoked by ``auth_user_passthru``. I wasn't sure if one occurrence would justify adding just another functionality to ``get_policies``.
What do you think?
- [x] <a href='#crh-comment-Pull e900fb6ace72202a32a2b541120a1deea8e0efdd privacyidea/lib/policy.py 42'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1062#discussion_r187925004'>File: privacyidea/lib/policy.py:L590-623</a></b>
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> I think this might often be a non-expected result. If there are two policies, with the same priority and the same resulting value - why should this raise an error?
The admin would argue, that the configuration create a clear logic with no contradiction. So he probalby would expect privacyIDEA to act accordingly.
- [x] <a href='#crh-comment-Pull e900fb6ace72202a32a2b541120a1deea8e0efdd privacyidea/lib/policydecorators.py 7'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1062#discussion_r187926517'>File: privacyidea/lib/policydecorators.py:L300-308</a></b>
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> Should we add this logic to ``get_policies``?
As we now have a solution for contradiction policies, there should be no contradicting policies anymore. So we could move the remaining contradicting stuff the the lower level of get_policies.
Depending on additional parameters like *sorted* and *unique* we could
1. return a sorted list, like for displaying in the webui.
2. return the top most priority policy if *unique* is set. And only if the logic is not clear, raise an exception in case of a failing *unique* from within ``get_policies``.
- [x] <a href='#crh-comment-Pull e900fb6ace72202a32a2b541120a1deea8e0efdd privacyidea/lib/policydecorators.py 17'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1062#discussion_r187926664'>File: privacyidea/lib/policydecorators.py:L358-376</a></b>
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> We then could also pass the *unique* to the get_policies in this function.
- [x] <a href='#crh-comment-Pull e900fb6ace72202a32a2b541120a1deea8e0efdd privacyidea/static/components/config/views/config.policies.details.html 17'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1062#discussion_r187928375'>File: privacyidea/static/components/config/views/config.policies.details.html:L322-342</a></b>
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> In the event handler list we also have an input field to change the "ordering" (priority) of the event handlers, which is actually used to determine, in which order the event handlers are applied.
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/static/components/config/views/config.events.list.html#L19
I would love to be able to also change the same/similar way the priority in the policy list. This way the admin could "sort" the policy list and move policies up and down.
- [x] <a href='#crh-comment-Pull 1b4377383a712e33c51815887e5afceda22e8059 privacyidea/lib/policy.py 57'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1062#discussion_r188183197'>File: privacyidea/lib/policy.py:L591-626</a></b>
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> Can we move this up? If there is only one policy we also do not need to create prioritizes policy...
So we could do in line 613:
if policies and len(policies) > 1:
- [x] <a href='#crh-comment-Pull 1b4377383a712e33c51815887e5afceda22e8059 privacyidea/lib/policy.py 107'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1062#discussion_r188183584'>File: privacyidea/lib/policy.py:L671-682</a></b>
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> I like this! :-)
- [x] <a href='#crh-comment-Pull 1b4377383a712e33c51815887e5afceda22e8059 privacyidea/lib/policydecorators.py 76'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1062#discussion_r188183959'>File: privacyidea/lib/policydecorators.py:L515-526</a></b>
- <a href='https://github.com/cornelinux'><img border=0 src='https://avatars1.githubusercontent.com/u/1908620?v=4' height=16 width=16></a> I think we can combine this to:
~~~~python
if login_mode_list and login_mode_list[0] == LOGINMODE:PRIVACYIDEA:


<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/1062?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/1062?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/privacyidea/privacyidea/pull/1062'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

cornelinux commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/privacyidea/privacyidea/pull/1062'>Refresh</a>.<a href='#crh-update-none'></a>

cornelinux commented:
Some thoughts on your comments before actually looking at the code:

> Initially I thought PolicyObject.get_policies should just return the policies sorted by priority. However, I decided against this because this would change the order of policies everywhere, e.g. also in the policy list WebUI. So instead, the priority field is only processed in situations where policy conflicts can occur.

Acutally changing the order everywhere might be quite nice, wouldn't it? This way the admin could also "sort" the policies in the webui, which would also help him to understand the application of the policies.

> This PR slightly changes behavior in some cases: In privacyIDEA 2.22, it is OK to have multiple matching policies otppin as long as they contain the same action value. With these changes, this would raise an error if they all have the same priority, even if they specify the same action value. Do you think we should keep backwards-compatibility here?

I think we should keep backwards-compatibility. *otppin* is I think fetched by ``get_action_value``. We could simply return the list of the highest priority action value. If there are more than one values/policies with different priority, we can still raise an exception.

cornelinux commented on privacyidea/lib/policy.py:
```diff
@@ -588,6 +590,33 @@ def get_policies(self, name=None, scope=None, realm=None, active=None,
+            total_priorities = len(priorities)
+            distinct_priorities = len(set(priorities))
+            if total_priorities != distinct_priorities:
+                raise PolicyError(u"There are policies with conflicting priorities")
```
I think this might often be a non-expected result. If there are two policies, with the same priority and the same resulting value - why should this raise an error?

The admin would argue, that the configuration create a clear logic with no contradiction. So he probalby would expect privacyIDEA to act accordingly.

cornelinux commented on privacyidea/lib/policydecorators.py:
```diff
@@ -300,8 +300,8 @@ def auth_user_passthru(wrapped_function, user_object, passw, options=None):
-            log.debug(u"Contradicting passthru policies: {0!s}".format(pass_thru))
-            raise PolicyError("Contradicting passthru policies.")
+            pass_thru = policy_object.choose_prioritized_policy(pass_thru)
+            log.debug(u"Multiple passthru policies, choosing {0!s}".format(pass_thru[0]))
```
Should we add this logic to ``get_policies``?

As we now have a solution for contradiction policies, there should be no contradicting policies anymore. So we could move the remaining contradicting stuff the the lower level of get_policies.

Depending on additional parameters like *sorted* and *unique* we could

1. return a sorted list, like for displaying in the webui.
2. return the top most priority policy if *unique* is set. And only if the logic is not clear, raise an exception in case of a failing *unique* from within ``get_policies``.

cornelinux commented on privacyidea/lib/policydecorators.py:
```diff
@@ -358,19 +358,18 @@ def auth_user_timelimit(wrapped_function, user_object, passw, options=None):
                                                       user=user_object.login,
-                                                      client=clientip)
+                                                      client=clientip,
+                                                      unique=True)
```
We then could also pass the *unique* to the get_policies in this function.

cornelinux commented on privacyidea/static/components/config/views/config.policies.details.html:
```diff
@@ -322,6 +322,20 @@
+            </p>
+        </div>
+    </div>
+
```
In the event handler list we also have an input field to change the "ordering" (priority) of the event handlers, which is actually used to determine, in which order the event handlers are applied.
https://github.com/privacyidea/privacyidea/blob/master/privacyidea/static/components/config/views/config.events.list.html#L19

I would love to be able to also change the same/similar way the priority in the policy list. This way the admin could "sort" the policy list and move policies up and down.

fredreichbier commented:
I added the priorities and the sorting functionality to the policy list and changed the conflict resolution as we discussed (see ``get_action_values``).

I also added the ``sort_by_priority`` flag to ``get_policies`` as discussed and set its default value to True. Thus, all policies are now sorted by priority. Doesn't seem to break anything. :)

I am hesitant to add the ``unique`` flag to ``get_policies``: Actually, it seems like ``auth_user_passthru`` is the only place where the action values are significant and ``get_policies`` is used instead of ``get_action_values``. We cannot just use ``get_action_values`` there because 1) ``get_action_values`` does not support boolean actions for now and 2) passthru needs the name of the matching policy. So instead I just added a conflict checking function to ``PolicyObject`` which is then invoked by ``auth_user_passthru``. I wasn't sure if one occurrence would justify adding just another functionality to ``get_policies``.
What do you think?

cornelinux commented on privacyidea/lib/policy.py:
```diff
@@ -586,8 +591,35 @@ def get_policies(self, name=None, scope=None, realm=None, active=None,
+        if policies:
+            prioritized_policy = min(policies, key=itemgetter("priority"))
+            prioritized_action = prioritized_policy["action"][action]
+            if len(policies) > 1:
```
Can we move this up? If there is only one policy we also do not need to create prioritizes policy...

So we could do in line 613:

   if policies and len(policies) > 1:

cornelinux commented on privacyidea/lib/policy.py:
```diff
@@ -635,10 +671,11 @@ def get_action_values(self, action, scope=SCOPE.AUTHZ, realm=None,
+        # Check if the policies with the highest priority agree on the action values
+        if unique and len(action_values) > 1:
+            names = [p['name'] for p in policies]
+            raise PolicyError(u"There are policies with conflicting actions: {!r}".format(names))
```
I like this! :-)

cornelinux commented on privacyidea/lib/policydecorators.py:
```diff
@@ -521,15 +515,11 @@ def login_mode(wrapped_function, *args, **kwds):
-                # exception
-                raise PolicyError(ERROR)
-
             if login_mode_list[0] == LOGINMODE.PRIVACYIDEA:
```
I think we can combine this to:

~~~~python
if login_mode_list and login_mode_list[0] == LOGINMODE:PRIVACYIDEA:
